charmcraft/charm_builder.py

```diff
@@ -0,0 +1,357 @@
+# Copyright 2020-2021 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+"""The charm package builder."""
+
+import argparse
+import errno
+import logging
+import os
+import pathlib
+import shutil
+import site
+import subprocess
+import sys
+from typing import List
+
+from charmcraft.cmdbase import CommandError
+from charmcraft.jujuignore import JujuIgnore, default_juju_ignore
+from charmcraft.utils import make_executable
+
+
+# Some constants that are used through the code.
+WORK_DIRNAME = "work_dir"
+VENV_DIRNAME = "venv"
+STAGING_VENV_DIRNAME = "staging-venv"
+
+# The file name and template for the dispatch script
+DISPATCH_FILENAME = "dispatch"
+# If Juju doesn't support the dispatch mechanism, it will execute the
+# hook, and we'd need sys.argv[0] to be the name of the hook but it's
+# geting lost by calling this dispatch, so we fake JUJU_DISPATCH_PATH
+# to be the value it would've otherwise been.
+DISPATCH_CONTENT = """#!/bin/sh
+
+JUJU_DISPATCH_PATH="${{JUJU_DISPATCH_PATH:-$0}}" PYTHONPATH=lib:venv ./{entrypoint_relative_path}
```

Comment:
This needs to be fixed if not using a "target" type of installation, but a virtualenv:

```
$ python3 -m venv foo
$ foo/bin/pip install requests
Collecting requests
  Using cached requests-2.26.0-py2.py3-none-any.whl (62 kB)
Collecting idna<4,>=2.5; python_version >= "3"
  Using cached idna-3.2-py3-none-any.whl (59 kB)
Collecting charset-normalizer~=2.0.0; python_version >= "3"
  Downloading charset_normalizer-2.0.3-py3-none-any.whl (35 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2021.5.30-py2.py3-none-any.whl (145 kB)
Collecting urllib3<1.27,>=1.21.1
  Using cached urllib3-1.26.6-py2.py3-none-any.whl (138 kB)
Installing collected packages: idna, charset-normalizer, certifi, urllib3, requests
Successfully installed certifi-2021.5.30 charset-normalizer-2.0.3 idna-3.2 requests-2.26.0 urllib3-1.26.6
$ PYTHONPATH=foo python3 -c "import requests;print(requests.__file__)"
/usr/lib/python3/dist-packages/requests/__init__.py
$ PYTHONPATH=foo/lib/python3.8/site-packages/ python3 -c "import requests;print(requests.__file__)"
/home/facundo/temp/tvenv/foo/lib/python3.8/site-packages/requests/__init__.py
```

Reply:
After the builder tool runs, only the virtual env `site.USER_SITE` directory is copied to the target `venv` directory, in such a way that the final charm payload remains with the same structure.